boot/freeldr/freeldr/arch/uefi/uefimem.c

```diff
@@ -0,0 +1,309 @@
+
+    TRACE("Value of MapKey: %d\n", MapKey);
+    TRACE("Value of MapSize: %d\n", MapSize);
+    TRACE("Value of DescriptorSize: %d\n", DescriptorSize);
+    TRACE("Value of DescriptorVersion: %d\n", DescriptorVersion);
+
+    EntryCount = (MapSize / DescriptorSize);
+
+    FreeldrMemMapSize = (sizeof(FREELDR_MEMORY_DESCRIPTOR) * EntryCount);
+	Status = GlobalSystemTable->BootServices->AllocatePool(EfiLoaderData, FreeldrMemMapSize, (void**)&FreeldrMem);
+    if (Status != EFI_SUCCESS)
+    {
+        TRACE("Failed to allocate pool with status %d", Status);
+        UiMessageBoxCritical("Unable to initialize memory manager.");
+    }
+
+    RtlZeroMemory(FreeldrMem, FreeldrMemMapSize);
+    MapEntry = EfiMemoryMap;
+	for (Index = 0; Index < EntryCount; ++Index)
+    {
+        if (UefiConvertToFreeldrDesc(MapEntry->Type) == LoaderFree)
+        {
+            Status = GlobalSystemTable->BootServices->AllocatePages(AllocateAddress, EfiLoaderData,
+                                                                    MapEntry->NumberOfPages, &MapEntry->PhysicalStart);
+            if (Status == EFI_SUCCESS)
+            {
+                UefiSetMemory(FreeldrMem, MapEntry->PhysicalStart, MapEntry->NumberOfPages, LoaderFree);
+            }
+            else
+            {
+                UefiSetMemory(FreeldrMem, MapEntry->PhysicalStart, MapEntry->NumberOfPages, LoaderFirmwareTemporary);
+            }
+        }
+        else
+        {
+            UefiSetMemory(FreeldrMem, MapEntry->PhysicalStart, MapEntry->NumberOfPages, UefiConvertToFreeldrDesc(MapEntry->Type));
+        }
+        MapEntry = NEXT_MEMORY_DESCRIPTOR(MapEntry, DescriptorSize);
+    }
+
+    *MemoryMapSize = FreeldrDescCount;
+    return FreeldrMem;
+}
+
+VOID
+UefiExitBootServices(VOID)
+{
+    UINTN MapKey;
+    UINTN MapSize;
+    EFI_STATUS Status;
+    UINTN DescriptorSize;
+    UINT32 DescriptorVersion;
+
+    TRACE("Attempting to exit bootsevices\n");
+    PUEFI_LoadMemoryMap(&MapKey,
+                        &MapSize,
+                        &DescriptorSize,
+                        &DescriptorVersion);
+
+    Status = GlobalSystemTable->BootServices->ExitBootServices(GlobalImageHandle, MapKey);
+    /* UEFI spec demands twice! */
+    if (Status != EFI_SUCCESS)
+        Status = GlobalSystemTable->BootServices->ExitBootServices(GlobalImageHandle, MapKey);
+
+    if (Status != EFI_SUCCESS)
+    {
+        TRACE("Failed to exit boot services with status: %d\n", Status);
+        FrLdrBugCheckWithMessage(EXIT_BOOTSERVICES_FAILURE,
+                                 __FILE__,
+                                 __LINE__,
+                                 "Failed to exit boot services with status: %d",
+                                 Status);
+    }
+    else
+    {
+        TRACE("Exited bootservices\n");
+    }
+}
+
+VOID
+UefiPrepareForReactOS(VOID)
+{
+    UefiExitBootServices();
+    ExitStack = MmAllocateMemoryWithType(EXIT_STACK_SIZE, LoaderOsloaderStack);
```

Comment:
Should this be done after, or before exiting the boot services?

Reply:
It shouldn’t matter but I prefer as much memory allocation to be done post exit as possible. This is just an organization choice, I doubt anything will care since every piece of memory freeldr uses gets specifically requested instead of just making all free memory LoaderFree.

Reply:
ok I was asking just to be sure.